### ISSUE TITLE
lint checker wrongly marks private method used in test as dead_code

### ISSUE THREAD

**Issue by ananthp** [rust-lang/rust#71990]
Consider this code:

```rust
struct Foo {
    data: Vec<usize>,
}

impl Foo {
    pub fn new() -> Foo {
        Foo {
            data: vec![1, 2, 3],
        }
    }

    // Warning: dead_code
    fn valuable_insight_only_used_in_test(&self) -> usize {
        self.data.len()
    }
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn makes_exactly_three_numbers_by_default() {
        let foo = Foo::new();
        assert_eq!(3, foo.valuable_insight_only_used_in_test());
    }
}
```

struct `Foo` has a private method `valuable_insight_only_used_in_test` that's not referred to in the rest of the struct's implementation, but is used in test. Compiler wrongly issues a `dead_code` lint warning.

```
...
  = note: `#[warn(dead_code)]` on by default
...
warning: associated function is never used: `valuable_insight_only_used_in_test`
  --> src/lib.rs:13:5
   |
13 |     fn valuable_insight_only_used_in_test(&self) -> usize {
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

Note: When this code is run on [Rust Playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=0b8cb094dbafc172a3bbeb6096c503da), the struct itself gets flagged as dead_code. However, in a real lib only the method triggers the warning.

**Comment by jonas-schievink** [comment]
This is expected behavior due to the use of `#[cfg(test)]`. When not building in test mode, the `test` module is missing, so the function is unused from the lint's perspective. You can annotate the function with `#[cfg(test)]` or silence the lint for it.
